### ISSUE TITLE
Off-by-one error in fee calculation

### ISSUE THREAD

**Issue by tcharding** [rust-bitcoin/rust-bitcoin#3806]
We use sat/kwu and Core uses sats/kvb. Likely because of div by 4 we have an off by one error when calculating the fee from an amount and fee rate. 

Question: Why did we choose kwu in `FeeRate` instead of kvb? Was it to match weight units in `Weight`?

### Proof 

These are part of the tests from Core

```c++
    feeRate = CFeeRate(123);
    // Rounds up the result, if not integer
    BOOST_CHECK_EQUAL(feeRate.GetFee(0), CAmount(0));
    BOOST_CHECK_EQUAL(feeRate.GetFee(8), CAmount(1)); // Special case: returns 1 instead of 0
    BOOST_CHECK_EQUAL(feeRate.GetFee(9), CAmount(2));
    BOOST_CHECK_EQUAL(feeRate.GetFee(121), CAmount(15));
    BOOST_CHECK_EQUAL(feeRate.GetFee(122), CAmount(16));
    BOOST_CHECK_EQUAL(feeRate.GetFee(999), CAmount(123));
    BOOST_CHECK_EQUAL(feeRate.GetFee(1e3), CAmount(123));
    BOOST_CHECK_EQUAL(feeRate.GetFee(9e3), CAmount(1107));
```

Which in Rust is this, and it fails with an off-by-one for 122 virtual bytes.
```rust
    #[test]
    fn core_get_fee() {
        // Bitcoin Core's GetFee function uses virtual bytes where
        // as we use weight units. Remember `1vb = 4wu`.
        let fee_rate = FeeRate::from_sat_per_kvb(123);

        assert_eq!(fee_rate.fee_vb(0).unwrap(), Amount::ZERO);
        assert_eq!(fee_rate.fee_vb(8).unwrap(), Amount::from_sat(1));
        assert_eq!(fee_rate.fee_vb(9).unwrap(), Amount::from_sat(2));
        assert_eq!(fee_rate.fee_vb(121).unwrap(), Amount::from_sat(15));
        assert_eq!(fee_rate.fee_vb(122).unwrap(), Amount::from_sat(16));
        assert_eq!(fee_rate.fee_vb(999).unwrap(), Amount::from_sat(123));
        assert_eq!(fee_rate.fee_vb(1_000).unwrap(), Amount::from_sat(123));
        assert_eq!(fee_rate.fee_vb(9_000).unwrap(), Amount::from_sat(123));
    }
```
 
I wrote this issue instead of just fixing it to discuss the question and make sure we are happy with it.

To close this issue I would expect a bucket load of unit tests be added to the `units` module. (And likely that they are taken from Core but writing bugs in the unit tests will be easy to do because of the whole kvb vs kwu thing.)

**Comment by sanket1729** [comment]
> Why did we choose kwu in FeeRate instead of kvb? Was it to match weight units in Weight?

I always use sats/kvb or sats/vb. I can guess we opted for this to maintain maximum precision and to support both from_sat_per_kwu and from_sats_per_pvu. If we were to define FeeRate(u64) in terms of kvb, it would result in a loss of precision when using the from_sats_per_kwu constructor?

**Comment by tcharding** [comment]
I know what you mean :)

**Comment by tcharding** [comment]
Note that currently we have undocumented loss of precision in `FeeRate::from_sat_per_kvb`

**Comment by apoelstra** [comment]
> Why did we choose kwu in `FeeRate` instead of kvb? Was it to match weight units in `Weight`?

Yes, weight is a more natural unit than vbytes, which is inherently imprecise. But unfortunately "per vbyte" is more precise than "per wu".

> Note that currently we have undocumented loss of precision in `FeeRate::from_sat_per_kvb`

Yeah, I see. This shouldn't happen.

So, we have a couple options:

* Just live with it
* Change `from_sat_per_kvb` to round up; or split it into _ceil/_floor methods.
* Change the internal repr of `FeeRate` to be fee/kwu to match Core (and do no better)
* Change the internal repr of `FeeRate` to use a f64 (which will handle integers <= MAX_MONEY and multiplication/division by 4 perfectly)

BTW this whole `from_sat_per` collection of functions is a bit weird looking. As an independent thing, maybe we want to replace/augment these with `FeeRate::per_*` methods that just take an `Amount`?

**Comment by apoelstra** [comment]
After reading https://github.com/rust-bitcoin/rust-bitcoin/issues/3807 I am warming up to the f64 solution.

We can add an `to_sat_f64` method to `Amount`/`SignedAmount`/`Weight` to avoid casting, and `Weight::from_f64_floor` and `Weight::from_f64_ceil`. Then we taboo `/ 4` and insist that it be `/ 4.0` (or have a justifying comment). Then all of our conversions will be exact and any loss of precision is made explicit by the use of `_floor` or `_ceil` functions.

(In the case of the code quoted in #3807 we might just keep the existing "107 / 4" but we should comment that this is a deliberate truncation and was copied from Core.)

**Comment by yancyribbens** [comment]
Checked that `123 sat/kvb * 122 vb = 15.006 sats`.  Rounding up would give 16.

> Change the internal repr of FeeRate to be fee/kwu to match Core (and do no better)

did you mean `sats/kvb` since the issue states that's what is used as the basis for core?

> Change the internal repr of FeeRate to use a f64

I thought for financial application, it's preferable to use decimal over float.  I think it follows that if decimal/float is used for the back-end calculation, there are much less precision rounding errors that could be.

**Comment by yancyribbens** [comment]
> I think that it's fine for library authors to use IEEE 754 floats based on a correct understanding of them instead of blindly applying generic rules of thumb.

Agree to may as well use what's easiest to begin with until there's an actual reason to add complexity.  Probably easiest to go with IEEE 754 float.

**Comment by yancyribbens** [comment]
> We use sat/kwu and Core uses sats/kvb

In Core, if I call ToString of a FeeRate and print the results, I get:

`feerate: 0.00005000 BTC/kvB`.  Note, that this prints BTC/kvB not sats/kvb.  I would prefer if this was just printing sats/kvB, not sure why the decision to print the results as BTC/kvB.

**Comment by yancyribbens** [comment]
A possible fix for this would be to change the internal representation of FeeRate to be a fraction.  This is currently being added to core: https://github.com/bitcoin/bitcoin/pull/30535 and may replace the corresponding core `CFeeRate` type.

**Comment by apoelstra** [comment]
Innnteresting. In principle I like this, and was only hesitant to propose it because Core was doing different things and I think there's value in matching Core exactly for feerate calculations. So if Core is moving that direction then it becomes more appealing.

But. It is a pretty complicated change that will rarely be useful. If we just change our internal representation from fee/kwu to fee/kvb, i.e. making it a "multiple of 0.25" instead of "multiple of 1", that will be equivalent for all feerates that are directly constructible in this library. Where they differ is: what happens when you add two feerates together, or multiply/divide one by an integer.

Seems like a lot of fiddly error-prone work (and perf regressions, and also tons of new overflow cases to think about), and may not be worth it.

Probably we just want to "reserve the right" to mess around with precision or overflow behavior post-1.0 and leave it be as-is. And to fix this specific issue I suggest we just do a one-time addition the extra 2 bits of precision.

**Comment by jsarenik** [comment]
A simple thing is to add three. Example:

```sh
weight=146760
echo minimal fee in absolute sats also known as vsize is $(( ($weight+3)/4 ))
```

**Comment by Kixunil** [comment]
The reason we have per kwu now is because when I implemented it either Core had the same type or I missed the difference. Anyway, concept ACK changing it to per kvb because it's both more precise and what Core does.
